playground/attention_with_rpe_template.py

```diff
@@ -0,0 +1,221 @@
+            (0, True)
+        )
+        offset_mapping = tkw.IndexMapping(
+            num_iterators=2,
+            inputs={
+                M: i,
+                K2: clip_lower_and_upper
+            },
+            outputs={ M: i, K2: j },
+            dynamic_val_mappings=({ M: i, K2: j }, { M: i, K2: j })
+        )
+
+        rpe_layout = tkl.MemoryLayout(shape=[max_context_length, ])
+    else:
+        rpe_layout = tkl.MemoryLayout(shape=[1, ])
+    # yapf: enable
+
+    @tkw.wave(constraints)
+    def base_attention(
+        q: tkl.Memory[B, M, K1, GLOBAL_ADDRESS_SPACE, tkl.f16],
+        k: tkl.Memory[B, K2, K1, ADDRESS_SPACE, tkl.f16],
+        v: tkl.Memory[B, N, K2, ADDRESS_SPACE, tkl.f16],
+        # TODO: if not use_t5_rpe, this will DCE; atm DCE on blockargs crashes.
+        rpe: tkl.Memory[K2, GLOBAL_ADDRESS_SPACE, tkl.f32, rpe_layout],
+        c: tkl.Memory[B, M, N, GLOBAL_ADDRESS_SPACE, tkl.f32],
+        debug_out: tkl.Memory[B, M, K2, GLOBAL_ADDRESS_SPACE, tkl.f32],
+    ):
+        c_reg = tkl.Register[B, N, M, tkl.f32](0.0)
+        init_sum = tkl.Register[B, M, tkl.f32](0.0)
+        init_max = tkl.Register[B, M, tkl.f32](-1e6)
+
+        # This microkernel encodes the fact that if the reduction
+        # dimension were tiled, then we would need to materialize a loop.
+        @tkw.reduction(K2, init_args=[init_max, init_sum, c_reg])
+        def repeat(
+            partial_max: tkl.Register[B, M, tkl.f32],
+            partial_sum: tkl.Register[B, M, tkl.f32],
+            acc: tkl.Register[B, N, M, tkl.f32],
+        ):
+            imm_reg = tkl.Register[B, K2, M, tkl.f32](0.0)
+            q_reg = tkw.read(q, elements_per_thread=LOAD_ELEMS_PER_THREAD_QK)
+            k_reg = tkw.read(k, elements_per_thread=LOAD_ELEMS_PER_THREAD_QK)
+            inner_acc = tkw.mma(k_reg, q_reg, imm_reg, mfma_variant[0])
+            x_j = tkw.permute(inner_acc, target_shape=[B, M, K2])
+
+            ####################################################################
+            # T5 RPE
+            ####################################################################
+            # Fused T5 RPE adds attention bias pre-softmax normalization.
+            # When fusing into the FA variant, adding locally before the max and
+            # the partial softmax should be equivalent.
+            if use_t5_rpe:
+                i = tkw.self_index(M, tkl.i64, elements_per_thread=1)
+                i = tkw.broadcast(i, target_shape=[B, M, K2])
+                j = tkw.self_index(
+                    K2, tkl.i64, elements_per_thread=LOAD_ELEMS_PER_THREAD_QK)
+
+                # yapf: disable
+                rpe_reg = tkw.read(
+                    rpe,
+                    mapping=offset_mapping,
+                    mapping_dynamic_vals=( i, j, ),
+                    elements_per_thread=LOAD_ELEMS_PER_THREAD_QK,
+                )
+                # yapf: enable
+                tkw.write(rpe_reg,
```

Comment:
Is this write only for debugging purposes? If so, can you remove?

Reply:
ok, but note I used it to test the RPE cond matrix itself is consistent, which let me find the missing log2e factor.